controller/service.go

```diff
@@ -179,3 +183,105 @@ func (c *controller) allocateIP(key string, svc *v1.Service) (net.IP, error) {
 	// Okay, in that case just bruteforce across all pools.
 	return c.ips.Allocate(key, isIPv6, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc))
 }
+
+// Dual-stack
+
+func (c *controller) convergeBalancerDual(l log.Logger, key string, svc *v1.Service) bool {
```

Comment:
From comments above;

> I have tried quite hard to keep the dual-stack code separated from the old code. New functions rather than alter the parameters and interleave new code with the old.

From experience this is what maintainers wants. The main concern is to not break anything old that works. But if you have trust in your regression tests, I can certainly expore the possibility. But be aware that "if single-stack; then ... else" all over the place is _not_ an improvment or easier to maintain, so if it goes that way I will back out, or suggest a larger re-factoring.

Reply:
Another reason for keeping single and dual stack code separated is that singe-stack cluster will become obsolete when dual-stack becomed GA. That is, you can't "turn-off" dual-stack even if only one family is actually used. Then it is nice to remove obsolete code from metallb rather than maintain it forever.

Reply:
> From experience this is what maintainers wants. The main concern is to not break anything old that works. 

I prefer clean code and pay the price to handle multiple ips when adding the dual stack seems reasonable IMHO.

> But if you have trust in your regression tests, I can certainly expore the possibility. But be aware that "if single-stack; then ... else" all over the place is _not_ an improvment or easier to maintain, so if it goes that way I will back out, or suggest a larger re-factoring.

Yes, definitely, I agree. I don't want an "if dual ... then ..." either :+1: 

> Another reason for keeping single and dual stack code separated is that singe-stack cluster will become obsolete when dual-stack becomed GA. That is, you can't "turn-off" dual-stack even if only one family is actually used. Then it is nice to remove obsolete code from metallb rather than maintain it forever

True, that is a good point. The single stack can be removed in a year or two, it is still quite some time. This PR, though, is kind of hacky with "ip" and "ip2" instead of a `ips` slice or something. And if we write using something using a slice to handle dual stack, won't it be very close (if not just "working") to handle also the single stack? What do you think?

Reply:
+1 on considering to unify the logic. The two functions do not seem to far if we had allocate always returning a slice. A lot of logic is done on the first element anyway.